### PR TITLE
feat: Android Global Context

### PR DESCRIPTION
**Pull request by carlossteinzup**
## Description

The **Global Context** was created using an **Object** class in kotlin since we need it to be available while the application is running.

A Global context is a singleton **ContextData** object that could be called anywhere in the application. 

it could be called in Declative screens (either in Front-end Android or full Server-Driven).

**Details**
A Global Context will **set** itself in a local context as well, with _ID = "global"_, to use the ContextDataManager orchestration and vice versa. It will update itself into the Object instance when a context with _ID = "global"_ is set locally.

To **get** it in a Server-driven screen use the **@{global}** walking in the JSON tree according to the _"paths"_ used when it was set.
To **set** it in a Serven-driven screen use the **SetContext** function listing its ID as _"global"_

To **get** it **programmatically** in any native screen use **`Global.get()`**. 
- A path could be passed on this function to get a specific path inside a JSON object

To **set** it **programmatically** in any native screen use **`Global.set(value: Any)`**. 
- A path could be passed on this function to get a specific path inside a JSON object. 
- A value **MUST** be passed on this function to set a specific value. This value **could** be an empty string

To **clear** it programmatically in any native screen use **`Global.clear()`**. 

- A path could be passed on this function to clear a specific path inside a JSON object

## Related Issues
https://github.com/ZupIT/beagle/issues/538
https://github.com/ZupIT/beagle/issues/123

**Issue 1:** It has been noticed that when attempting to set a value for a context using the SetContext method in a Declarative screen, the context value only evaluates single strings information. It is not able to convert Objects into JSON objects through the set context.

**Issue 2:** The Clear method has been configured in a way to use the set context method to change the global context value to an empty string. Since it used the set mechanism, if a user lists a path that does not exist, it will create a new path, and it should return null when a path does not exist. 

We have left these 2 issued unsolved for the moment since a significative **CHANGE** will occur at the ContextDataManager file, which will change the way we will handle the set and clear methods. 

## Tests

I added the following tests:

    get_should_return_context_root_when_no_path_is_listed()
    get_should_return_context_with_path() 
    get_should_return_null_when_path_does_not_exist()
    set_should_set_global_context_value_at_root_when_path_is_null()
    set_should_set_global_context_value_at_a_specific_path() 
    set_should_not_override_other_paths_in_global_context_root() {
    clear_should_set_empty_string_value_at_global_context_root_when_path_is_null() {
    clear_sould_set_empty_string_value_at_global_context_with_path() {

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [DCO].
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

[issue database]: https://github.com/ZupIT/beagle/issues
[DCO]: https://developercertificate.org/

### CODE REVIEW

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">The title of PR not follow the rules of: https://www.conventionalcommits.org/en/v1.0.0/</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: android/beagle/src/main/java/br/com/zup/beagle/android/context/GlobalContext.kt [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+typealias GlobalContextObserver = (SetContext) -> Unit
+
+object GlobalContext {
+
+    private var globalContext = createGlobalContext()
+    private val globalContextObservers = mutableListOf<GlobalContextObserver>()
+    private val contextDataEvaluation = ContextDataEvaluation()
+    private val contextDataManager = ContextDataManager()
+
+    // Internal methods
```
Comment (ze12augusto) [inline android/beagle/src/main/java/br/com/zup/beagle/android/context/GlobalContext.kt:31]: You could remove these comments and use only spacing to separate the functions

File: android/beagle/src/test/java/br/com/zup/beagle/android/context/ContextDataTreeHelperTest.kt [hunk 2]
```diff
@@ -16,15 +16,13 @@
...
-import androidx.core.util.lruCache
-import br.com.zup.beagle.android.action.SetContextInternal
+import androidx.collection.LruCache
 import br.com.zup.beagle.android.jsonpath.JsonCreateTree
 import br.com.zup.beagle.android.testutil.RandomData
 import org.json.JSONArray
 import org.json.JSONObject
 import org.junit.Test
-import java.util.LinkedList
+import java.util.*
```
Comment (ze12augusto) [inline android/beagle/src/test/java/br/com/zup/beagle/android/context/ContextDataTreeHelperTest.kt:25]: Remove wildcard import

File: android/beagle/src/test/java/br/com/zup/beagle/android/context/GlobalContextTest.kt [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+class GlobalContextTest : BaseTest() {
+
+    override fun setUp() {
+        super.setUp()
+        GlobalContext.clearContext()
+    }
+
+    @Test
+    fun get_should_return_context_root_when_no_path_is_listed() {
+        // Given When
```
Comment (ze12augusto) [inline android/beagle/src/test/java/br/com/zup/beagle/android/context/GlobalContextTest.kt:34]: You could remove these comments and use only spacing to separate the lines

File: android/beagle/src/main/java/br/com/zup/beagle/android/context/GlobalContext.kt (hunk 1 above)
Comment (carlossteinzup) [reply android/beagle/src/main/java/br/com/zup/beagle/android/context/GlobalContext.kt]: Removed

File: android/beagle/src/test/java/br/com/zup/beagle/android/context/ContextDataTreeHelperTest.kt (hunk 2 above)
Comment (carlossteinzup) [reply android/beagle/src/test/java/br/com/zup/beagle/android/context/ContextDataTreeHelperTest.kt]: Removed

File: android/beagle/src/test/java/br/com/zup/beagle/android/context/GlobalContextTest.kt (hunk 3 above)
Comment (carlossteinzup) [reply android/beagle/src/test/java/br/com/zup/beagle/android/context/GlobalContextTest.kt]: Removed

File: android/beagle/src/main/java/br/com/zup/beagle/android/jsonpath/JsonCreateTree.kt [hunk 4]
```diff
@@ -22,20 +22,34 @@ import java.util.LinkedList
...
+        newValue: Any?,
+        createPathIfDoesNotExist: Boolean = true
+    ) {
         @Suppress("UNCHECKED_CAST")
         val copyOfKeys = keys.clone() as LinkedList<String>
         var key = copyOfKeys.poll()
 
-
         var currentTree: Any = root
         while (key != null) {
```
Comment (botbeaglezup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/jsonpath/JsonCreateTree.kt:36]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/dc39ea5ea2652ac8e225933e4b0350e0ce7901fa/android/beagle/src/main/java/br/com/zup/beagle/android/jsonpath/JsonCreateTree.kt#L36"></span>The loop contains more than one break or continue statement. The code should be refactored to increase readability.</td>
    </tr>
  </tbody>
</table>

File: android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManager.kt [hunk 5]
```diff
@@ -30,21 +29,34 @@ import br.com.zup.beagle.android.utils.setContextBinding
...
 
+    private var globalContext: ContextBinding = ContextBinding(GlobalContext.getContext())
     private val contexts = mutableMapOf<Int, ContextBinding>()
     private val viewBinding = mutableMapOf<View, MutableSet<Binding<*>>>()
+    private val globalContextObserver: GlobalContextObserver = {
+        updateGlobalContext(it)
+    }
+
+    init {
+        contexts[Int.MAX_VALUE] = globalContext
```
Comment (uziasferreirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManager.kt:44]: What do you think to change this constant to a variable to more readable something like 
```suggestion
           const val GLOBAL_CONTEXT_ID = Int.MAX_VALUE
               contexts[GLOBAL_CONTEXT_ID] = globalContext
```

File: android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManager.kt [hunk 6]
```diff
@@ -133,29 +152,15 @@ internal class ContextDataManager(
...
-            }
+    private fun View.getAllParentContextWithGlobal(): MutableMap<String, ContextBinding> {
+        return this.getAllParentContexts().apply {
+            put(globalContext.context.id, globalContext)
         }
     }
+
+    private fun updateGlobalContext(contextData: ContextData) {
+        globalContext = globalContext.copy(context = contextData)
+        contexts[Int.MAX_VALUE] = globalContext
```
Comment (uziasferreirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManager.kt:163]: the same suggestion
```suggestion
        contexts[GLOBAL_CONTEXT_ID] = globalContext
```

File: android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManipulator.kt [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+package br.com.zup.beagle.android.context
+
+import br.com.zup.beagle.android.jsonpath.JsonCreateTree
+import br.com.zup.beagle.android.jsonpath.JsonPathFinder
+import br.com.zup.beagle.android.jsonpath.JsonPathUtils
+import br.com.zup.beagle.android.logger.BeagleMessageLogs
+import java.util.LinkedList
+
+internal sealed class ContextSetResult {
+    class Succeed(val newContext: ContextData) : ContextSetResult()
```
Comment (uziasferreirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManipulator.kt:26]: What do you think about using `data class` here?

File: android/beagle/src/test/java/br/com/zup/beagle/android/context/ContextDataManagerTest.kt [hunk 8]
```diff
@@ -197,25 +182,24 @@ class ContextDataManagerTest : BaseTest() {
...
-    fun updateContext_should_return_false_when_contextId_does_not_exist() {
+    fun updateContext_should_call_global_context_when_id_is_global() {
         // Given
         val updateContext = SetContextInternal(RandomData.string(), false, null)
 
         // When
         val result = contextDataManager.updateContext(viewContext, updateContext)
 
         // Then
-        assertFalse(result)
```
Comment (uziasferreirazup) [inline android/beagle/src/test/java/br/com/zup/beagle/android/context/ContextDataManagerTest.kt:218]: is it correct to remove this assert? looks not has any assert

Comment (tuliopereirazup) [reply android/beagle/src/test/java/br/com/zup/beagle/android/context/ContextDataManagerTest.kt:218]: No. I really miss this line. Thanks.

**Pull request merged**